### PR TITLE
Linker to Kotlin

### PR DESCRIPTION
**Pull request by oldergod**

### CODE REVIEW

File: wire-schema/src/main/java/com/squareup/wire/schema/FileLinker.java [hunk 1]
```diff
@@ -64,7 +64,7 @@ ProtoFile protoFile() {
   private void addImportsRecursive(Set<String> sink, Collection<String> paths) {
     for (String path : paths) {
       if (sink.add(path)) {
-        FileLinker fileLinker = linker.getFileLinker(path);
+        FileLinker fileLinker = linker.getFileLinker$wire_schema(path);
```
Comment (oldergod) [inline wire-schema/src/main/java/com/squareup/wire/schema/FileLinker.java]: Those weirdess will go away when FileLinker becomes Kotlin too.

Comment (swankjesse) [reply wire-schema/src/main/java/com/squareup/wire/schema/FileLinker.java]: such weirdness

File: wire-schema/src/main/java/com/squareup/wire/schema/Linker.java [hunk 2]
```diff
@@ -13,448 +13,445 @@
...
+  private val loader: Loader
+  private val fileLinkers: MutableMap<String, FileLinker>
+  private val protoTypeNames: MutableMap<String, Type>
+  private val errors: MutableList<String>
+  private val contextStack: List<Any>
+  private val requestedTypes: MutableSet<ProtoType?>
+
+  internal constructor(loader: Loader) {
+    this.loader = loader
+    fileLinkers = LinkedHashMap()
```
Comment (swankjesse) [inline wire-schema/src/main/java/com/squareup/wire/schema/Linker.java]: `= mutableMap()`  ?

File: wire-schema/src/main/java/com/squareup/wire/schema/Linker.java [hunk 3]
```diff
@@ -13,448 +13,445 @@
...
-    this.fileLinkers = enclosing.fileLinkers;
-    this.protoTypeNames = enclosing.protoTypeNames;
-    this.contextStack = Util.concatenate(enclosing.contextStack, additionalContext);
-    this.errors = enclosing.errors;
-    this.requestedTypes = enclosing.requestedTypes;
+  private constructor(enclosing: Linker, additionalContext: Any) {
+    loader = enclosing.loader
+    fileLinkers = enclosing.fileLinkers
+    protoTypeNames = enclosing.protoTypeNames
+    contextStack =
```
Comment (swankjesse) [inline wire-schema/src/main/java/com/squareup/wire/schema/Linker.java]: nit: unwrap?

File: wire-schema/src/main/java/com/squareup/wire/schema/Linker.java [hunk 4]
```diff
@@ -13,448 +13,445 @@
...
   }
 
-  /** Returns the field named {@code field} on the message type of {@code self}. */
-  Field dereference(Field self, String field) {
+  /** Returns the field named `field` on the message type of `self`. */
+  fun dereference(
+    self: Field,
+    field: String
+  ): Field? {
+    @Suppress("NAME_SHADOWING") var field = field
```
Comment (swankjesse) [inline wire-schema/src/main/java/com/squareup/wire/schema/Linker.java]: We should just put this on the entire project. I disagree with this warning

File: wire-schema/src/main/java/com/squareup/wire/schema/Linker.java [hunk 5]
```diff
@@ -13,448 +13,445 @@
...
-        StringBuilder error = new StringBuilder();
-        String constant = entry.getKey();
-        int index = 1;
-        error.append(String.format("multiple enums share constant %s:", constant));
-        for (EnumType enumType : entry.getValue()) {
-          error.append(String.format("\n  %s. %s.%s (%s)",
-              index++, enumType.getType(), constant, enumType.constant(constant).getLocation()));
+    for ((constant, value) in nameToType.asMap()) {
+      if (value.size > 1) {
+        val error = buildString {
```
Comment (swankjesse) [inline wire-schema/src/main/java/com/squareup/wire/schema/Linker.java]: nice

**Pull request merged**